compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/fold_dma_waits.mlir

```diff
@@ -0,0 +1,230 @@
+  func.func @fold_dma_waits_same_bd_id() {
+    %c0 = arith.constant 0 : index
+    %c1 = arith.constant 1 : index
+    amdaie.workgroup {
+      %tile = amdaie.tile(%c0, %c1)
+      %tile_0 = amdaie.tile(%c0, %c0)
+      %buffer = amdaie.buffer(%tile) : memref<2048xi32, 1 : i32>
+      %buffer_1 = amdaie.buffer(%tile) : memref<2048xi32, 1 : i32>
+      %lock = amdaie.lock(%tile(4), 4)
+      %lock_2 = amdaie.lock(%tile(5), 0)
+      %0 = amdaie.logicalobjectfifo.from_buffers({%buffer, %buffer_1}, {%lock}, {%lock_2}) : memref<2048xi32, 1 : i32>, memref<2048xi32, 1 : i32> -> !amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>
+      %1 = hal.interface.binding.subspan layout(#pipeline_layout) binding(0) alignment(64) offset(%c0) flags("ReadOnly|Indirect") : memref<64x32xi32>
+      %2 = amdaie.logicalobjectfifo.placeholder{%tile_0} : !amdaie.logicalobjectfifo<memref<64x32xi32>>
+      %channel = amdaie.channel(%tile_0, 0, port_type = DMA, direction = MM2S)
+      %channel_3 = amdaie.channel(%tile, 0, port_type = DMA, direction = S2MM)
+      %3 = amdaie.flow({%channel} -> {%channel_3}) {is_packet_flow = false}
+      %4 = amdaie.connection(%0 {%channel_3}, %2 {%channel}, flow = %3) {connection_type = #amdaie<connection_type Packet>} : (!amdaie.logicalobjectfifo<memref<2048xi32, 1 : i32>, 2>, !amdaie.logicalobjectfifo<memref<64x32xi32>>)
+      amdaie.controlcode {
+        %5 = amdaie.logicalobjectfifo.from_memref %1, {%tile_0} : memref<64x32xi32> -> !amdaie.logicalobjectfifo<memref<2048xi32>>
+        memref.assume_alignment %1, 64 : memref<64x32xi32>
+        %bd_id = amdaie.bd_id(%tile_0, %c0)
+        %6 = amdaie.npu.half_dma_cpy_nd async %4(%5 [] [] [] bd_id = %bd_id channel = %channel) : !amdaie.logicalobjectfifo<memref<2048xi32>>
+        amdaie.npu.dma_wait(%6 : !amdaie.async_token)
+        %7 = amdaie.npu.half_dma_cpy_nd async %4(%5 [] [] [] bd_id = %bd_id channel = %channel) : !amdaie.logicalobjectfifo<memref<2048xi32>>
+        amdaie.npu.dma_wait(%7 : !amdaie.async_token)
+        amdaie.end
+      }
+    }
+    return
+  }
+}
+
+// -----
+
+// DMA queue has a maximum size of 4. To optimize, starting from 
+// the end of the control code, retain every 4th DMA wait operation 
+// while folding the others.
+// CHECK-LABEL: @fold_dma_waits_max_queue_size
+// CHECK:       %[[OBJECT_FIFO_0:.+]] = amdaie.logicalobjectfifo.from_buffers
+// CHECK:       %[[CHANNEL_0:.+]] = amdaie.channel
+// CHECK:       %[[CHANNEL_1:.+]] = amdaie.channel
+// CHECK:       %[[CONNECTION:.+]] = amdaie.connection
+// CHECK:         %[[OBJECT_FIFO_1:.+]] = amdaie.logicalobjectfifo.from_memref
+// CHECK:         %[[BD_ID_0:.+]] = amdaie.bd_id
+// CHECK:         %[[TOKEN_0:.+]] = amdaie.npu.half_dma_cpy_nd async %[[CONNECTION]](%[[OBJECT_FIFO_1]] [] [] [] bd_id = %[[BD_ID_0]] channel = %[[CHANNEL_0]]) : !amdaie.logicalobjectfifo<memref<2048xi32>>
+// CHECK:         amdaie.npu.dma_wait(%[[TOKEN_0]] : !amdaie.async_token)
```

Comment:
Why do we have a wait on the first `dma_cpy_nd`?

I think we should expect:
```
dma_cpy_nd
dma_cpy_nd
dma_cpy_nd
%0 = dma_cpy_nd
dma_wait(%0)
```

Instead of:

```
%0 = dma_cpy_nd
dma_wait(%0)
dma_cpy_nd
dma_cpy_nd
%1 = dma_cpy_nd
dma_wait(%0)
```

Reply:
I am traversing controlcode in reverse order, and the example actually is:
```ll
%0 = dma_cpy_nd
dma_wait(%0)
dma_cpy_nd
dma_cpy_nd
dma_cpy_nd
%1 = dma_cpy_nd
dma_wait(%0)
```
there are four `dma_cpy_nd` ops between.

Reply:
Sorry, not only for `nb_dma_cpy_nd_ops <= 4`, but also in the example above, you're using two waits in what could be implemented with one.

Reply:
but if we don't have a wait on the first op, then we will have to do this:
```
dma_cpy_nd
dma_cpy_nd
dma_cpy_nd
dma_cpy_nd
%0 = dma_cpy_nd
dma_wait(%0)
%1 = dma_cpy_nd
dma_wait(%1)
```
we still have two waits, since anyway we need one wait at the end of controlcode?

Reply:
> we still have two waits, since anyway we need one wait at the end of controlcode?

That's not necessarily true, because typically the outputs depend on the inputs, so just waiting for the outputs is good. But it's good to keep the wait on the inputs for now because it's hard to **proof** that we're already waiting on a dependent output DMA, so we don't need to wait on this DMA.

Even if we keep the waits, there are cases in which you will generate too many waits:

```
%0 = dma_cpy_nd
dma_wait(%0)
dma_cpy_nd
dma_cpy_nd
dma_cpy_nd
%1 = dma_cpy_nd
dma_wait(%1)
%2 = dma_cpy_nd
dma_wait(%2)
```
Here, 3 waits are used, but it can be implemented with 2:
```
dma_cpy_nd
dma_cpy_nd
dma_cpy_nd
%0 = dma_cpy_nd
dma_wait(%0)
dma_cpy_nd
%1 = dma_cpy_nd
dma_wait(%1)
```

Reply:
actually running the above example with my current code, I get the following, which has two waits:
```ll
dma_cpy_nd
%0 = dma_cpy_nd
dma_wait(%0)
dma_cpy_nd
dma_cpy_nd
dma_cpy_nd
%1 = dma_cpy_nd
dma_wait(%1)
```
(one thing is all these dma_cpy_nd need diffent BD ID)

Reply:
Ok, that's good, but I still think this is non-intuitive output, leading to confusion, and it seems like an arbitrary choice unless I am missing something? So why do you feel strongly about returning it in this order?

Reply:
it is just making the logic simpler by traversing backward. Otherwise, I need to check if the current wait is the last one for each connection

Reply:
> Otherwise, I need to check if the current wait is the last one for each connection

You can traverse in forward order and keep track of every 4th and last DMA on a connection, then only keep those waits. I don't see how that's more complex?

Reply:
The reason I think this is important is that the output IR should be as one would intuitively expect. A lot of the times I have to debug issues by just reading and understanding the IR and non-intuitive output is no fun when you're doing that.

Reply:
I tried something, but then I realized that traversing in forward order complicates the management of BD IDs. When encountering a duplicate BD ID, we need to keep the wait for the last DMA whichused that BD ID. Also, all the erasion decisions between those two DMAs need to be updated accordingly.

Reply:
Ok, could you add documentation to the function on why you iterate in reverse order and an example to show expected output for one of these more quirky cases above?

Reply:
Thanks, added now